Always annotate outer/inner edges in the plan

hardikar opened the pull request:
This way, in the case that graphviz decides to swap the left/right children - at least it's clearly visible instead of a human having to check through big plans.

Also for future reference, the way to generate these images is:
```
cd src/test/regress
cat plan.txt | ./explain.pl -option dot | dot -Tsvg -o /tmp/plan.svg
```
And open the SVG file in a browser.

@d @karthijrk @hsyuan

For example, for the plan : 
```
$ cat plan.txt
                                                        QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=0.00..431.00 rows=1 width=8)
   ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..431.00 rows=1 width=8)
         ->  Aggregate  (cost=0.00..431.00 rows=1 width=8)
               ->  GroupAggregate  (cost=0.00..431.00 rows=1 width=4)
                     Group By: i, j, t
                     ->  Sort  (cost=0.00..431.00 rows=1 width=16)
                           Sort Key: i, j, t
                           ->  Sequence  (cost=0.00..431.00 rows=1 width=16)
                                 ->  Partition Selector for foo (dynamic scan id: 1)  (cost=10.00..100.00 rows=34 width=4)
                                       Partitions selected: 2 (out of 2)
                                 ->  Dynamic Table Scan on foo (dynamic scan id: 1)  (cost=0.00..431.00 rows=1 width=16)
 Settings:  optimizer=on
 Optimizer status: PQO version 2.27.0
(13 rows)

```

It'll now add the outer/inner labels to the edges : 
```
digraph plan1 {
rankdir=BT;
"2" -> "1";
"3" -> "2";
"4" -> "3";
"5" -> "4";
"6" -> "5";
"7" -> "6" [label="outer" ] ;
"8" -> "6" [label="inner" ] ;
"1" [shape=record, label="Aggregate", style=filled, color="#66c2a5", fillcolor="#66c2a5"];
"2" [shape=record, label="Gather Motion 3:1  (slice1; segments: 3)", style=filled, color="#fc8d62", fillcolor="#fc8d62"];
"3" [shape=record, label="Aggregate", style=filled, color="#fc8d62", fillcolor="#fc8d62"];
"4" [shape=record, label="GroupAggregate", style=filled, color="#fc8d62", fillcolor="#fc8d62"];
"5" [shape=record, label="Sort", style=filled, color="#fc8d62", fillcolor="#fc8d62"];
"6" [shape=record, label="Sequence", style=filled, color="#fc8d62", fillcolor="#fc8d62"];
"7" [shape=record, label="Partition Selector for foo (dynamic scan id: 1)", style=filled, color="#fc8d62", fillcolor="#fc8d62"];
"8" [shape=record, label="Dynamic Table Scan on foo (dynamic scan id: 1)", style=filled, color="#fc8d62", fillcolor="#fc8d62"];

}
```

This results to this SVG:

<img width="782" alt="screen shot 2017-04-26 at 11 36 06 am" src="https://cloud.githubusercontent.com/assets/2496677/25450704/99463464-2a74-11e7-8019-9cd77c666a56.png">

danielgustafsson commented on src/test/regress/explain.pm:
```diff
@@ -1601,13 +1600,13 @@ sub nestedloop_fixup
-        $kidlist[1]->{nested_loop_position} = "left";
-        $kidlist[0]->{nested_loop_position} = "right";
+        $kidlist[1]->{label} = "outer";
+        $kidlist[0]->{label} = "inner";
```
Nitpick. While not the fault of this diff, I would prefer if we set the `@kidlist` members in the same order in both blocks making it more visually obvious that we are shifting them. A quick glance now makes it seem like both blocks do the same thing.

hardikar replied:
Will do.

hsyuan commented:
I just have something come to my mind. What if a nestloop join has a initplan? Will it mark initplan as outer?

hardikar commented:
Do you have a plan/ query in mind? We can check real quick. 

> On Apr 28, 2017, at 12:49 PM, Haisheng Yuan <notifications@github.com> wrote:
> 
> I just have something come to my mind. What if a nestloop join has a initplan? Will it mark initplan as outer?
> 
> ―
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

hsyuan commented:
create table t1(a int); create table t2(b int); create table t3(c int);
insert into t2 select i from generate_series(1,10000) i; insert into t3
select i from generate_series(1,100000) i; hyuan=# explain select (select a
from t1), b, c from t2,t3; QUERY PLAN
-------------------------------------------------------------------------------------------
Gather Motion 3:1 (slice3; segments: 3) (cost=2.02..6336.07 rows=100000
width=9) -> Nested Loop (cost=2.02..6336.07 rows=33334 width=9) InitPlan
(slice4) -> Gather Motion 3:1 (slice2; segments: 3) (cost=0.00..1.01 rows=1
width=2) -> Seq Scan on t1 (cost=0.00..1.01 rows=1 width=2) -> Broadcast
Motion 3:3 (slice1; segments: 3) (cost=0.00..1.05 rows=1 width=4) -> Seq
Scan on t2 (cost=0.00..1.01 rows=1 width=4) -> Seq Scan on t3
(cost=0.00..1111.00 rows=33334 width=5) Settings: optimizer=off Optimizer
status: legacy query optimizer (10 rows)

~ ~ ~
Haisheng Yuan

On Fri, Apr 28, 2017 at 12:58 PM, Shreedhar Hardikar <
notifications@github.com> wrote:

> Do you have a plan/ query in mind? We can check real quick.
>
> > On Apr 28, 2017, at 12:49 PM, Haisheng Yuan <notifications@github.com>
> wrote:
> >
> > I just have something come to my mind. What if a nestloop join has a
> initplan? Will it mark initplan as outer?
> >
> > ―
> > You are receiving this because you authored the thread.
>
> > Reply to this email directly, view it on GitHub, or mute the thread.
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/2310#issuecomment-298092680>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AOpD2Q9NBmkNwAr9b5MXjjc39E0cgeBqks5r0kT2gaJpZM4NJRFT>
> .
>

hsyuan commented:
I just realized the plan was corrupted. Paste it again:
```sql
create table t1(a int);
create table t2(b int);
create table t3(c int);

insert into t2 select i from generate_series(1,10000) i;
insert into t3 select i from generate_series(1,100000) i;

explain select (select a from t1), b, c from t2,t3;
                                        QUERY PLAN
-------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice3; segments: 3)  (cost=2.02..6336.07 rows=100000 width=9)
   ->  Nested Loop  (cost=2.02..6336.07 rows=33334 width=9)
         InitPlan  (slice4)
           ->  Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..1.01 rows=1 width=2)
                 ->  Seq Scan on t1  (cost=0.00..1.01 rows=1 width=2)
         ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..1.05 rows=1 width=4)
               ->  Seq Scan on t2  (cost=0.00..1.01 rows=1 width=4)
         ->  Seq Scan on t3  (cost=0.00..1111.00 rows=33334 width=5)
 Settings:  optimizer=off
 Optimizer status: legacy query optimizer
(10 rows)
```

hardikar commented:
@hsyuan Thanks for that - you're right - it doesn't work for NLJ with InitPlan because it has "3" children ... Or HashJoin.. 

![screen shot 2017-05-03 at 4 15 57 pm](https://cloud.githubusercontent.com/assets/2496677/25685383/e0568514-301b-11e7-80e3-7a7a7033f285.png)

Let me see if I can fix that.

hsyuan commented:
But it is good that it doesn't display wrong label. I think it is OK.

ashwinstar commented:
@hardikar good discussions but seems to have stalled, can we take it to completion ?

hlinnaka commented:
Ping? Is this ready to be committed? What remains to be done?

pivotal-mike commented:
@hsyuan @ashwinstar I think we're good to merge this?

ashwinstar commented:
@hardikar or @hsyuan can let us know if its ready or not. I have no idea.

hsyuan commented:
The change doesn't annotate join when the join has an initplan. But it
doesn't hurt to merge.

~ ~ ~
Haisheng Yuan

On Tue, Oct 31, 2017 at 3:20 AM, Ashwin Agrawal <notifications@github.com>
wrote:

> @hardikar <https://github.com/hardikar> or @hsyuan
> <https://github.com/hsyuan> can let us know if its ready or not. I have
> no idea.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/2310#issuecomment-340555215>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AOpD2ffEuo8FtLftTRKKDNx4Mzm_oivNks5sxiFugaJpZM4NJRFT>
> .
>

hardikar commented:
It actually won't annotate anything with 3 children. I'm afraid it'll annotate the InitPlan as outer in case there is an InitPlan with another child because it only cares about the number of children and not about what the children are.

hardikar commented:
Ok. Updated the script to handle InitPlans. Also fixed an issue when it was not recognizing InitPlans not in a separate slice.

So at this point this PR is complete! @hsyuan @hlinnaka @danielgustafsson, it would be great if you can review it once again real quick, and I can push it in soon.

Example outputs after the changes:
```
                                              QUERY PLAN
-------------------------------------------------------------------------------------------------------
 Result  (cost=0.03..0.05 rows=1 width=4)
   One-Time Filter: $0
   InitPlan  (slice2)
     ->  Limit  (cost=0.00..0.03 rows=1 width=0)
           ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..0.03 rows=1 width=0)
                 ->  Limit  (cost=0.00..0.01 rows=1 width=0)
                       ->  Result  (cost=0.00..2883.00 rows=86100 width=0)
                             ->  Append  (cost=0.00..2883.00 rows=86100 width=0)
                                   ->  Seq Scan on r_1_prt_1 r  (cost=0.00..961.00 rows=28700 width=0)
                                   ->  Seq Scan on r_1_prt_2 r  (cost=0.00..961.00 rows=28700 width=0)
                                   ->  Seq Scan on r_1_prt_3 r  (cost=0.00..961.00 rows=28700 width=0)
   ->  Result  (cost=0.00..0.01 rows=1 width=0)
 Settings:  optimizer=off
 Optimizer status: legacy query optimizer
(14 rows)
```

<img width="663" alt="screen shot 2017-10-31 at 4 28 39 pm" src="https://user-images.githubusercontent.com/2496677/32220669-9589f6e4-be58-11e7-8320-b653fe27c2b0.png">

```
                                        QUERY PLAN
-------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice3; segments: 3)  (cost=2.02..6336.07 rows=100000 width=9)
   ->  Nested Loop  (cost=2.02..6336.07 rows=33334 width=9)
         InitPlan  (slice4)
           ->  Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..1.01 rows=1 width=2)
                 ->  Seq Scan on t1  (cost=0.00..1.01 rows=1 width=2)
         ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..1.05 rows=1 width=4)
               ->  Seq Scan on t2  (cost=0.00..1.01 rows=1 width=4)
         ->  Seq Scan on t3  (cost=0.00..1111.00 rows=33334 width=5)
 Settings:  optimizer=off
 Optimizer status: legacy query optimizer
(10 rows)
```
<img width="778" alt="screen shot 2017-10-31 at 4 29 43 pm" src="https://user-images.githubusercontent.com/2496677/32220698-ba45f2c6-be58-11e7-845b-2ae0636d3b5d.png">

```
shardikar=# explain (select 1) union (select 2);
                         QUERY PLAN
------------------------------------------------------------
 Unique  (cost=0.05..0.06 rows=1 width=0)
   Group By: "outer".?column?
   ->  Sort  (cost=0.05..0.06 rows=1 width=0)
         Sort Key (Distinct): "outer".?column?
         ->  Append  (cost=0.00..0.04 rows=1 width=0)
               ->  Result  (cost=0.00..0.01 rows=1 width=0)
               ->  Result  (cost=0.00..0.01 rows=1 width=0)
 Settings:  optimizer=off
 Optimizer status: legacy query optimizer
(9 rows)
```

<img width="182" alt="screen shot 2017-10-31 at 4 30 37 pm" src="https://user-images.githubusercontent.com/2496677/32220735-da3062ec-be58-11e7-8b50-816b8b400e25.png">

danielgustafsson commented on src/test/regress/explain.pm:
```diff
@@ -1577,39 +1577,49 @@ sub human_num
         {
-            push @kidlist, $kid;
+            # Ignore InitPlans when deciding inner/outer child
+            if ($kid->{txt} !~ 'InitPlan')
```
Since this is a regexp and not a string comparison, I would prefer // instead of '' to make it clearer for anyone not used to reading Perl: `if ($kid->{txt} !~ /InitPlan/)`

hsyuan commented:
LGTM
